bundles/org.openhab.binding.withings/src/main/java/org/openhab/binding/withings/internal/api/AbstractAPIHandler.java

```diff
@@ -0,0 +1,105 @@
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.withings.internal.api;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.FormContentProvider;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.util.Fields;
+import org.openhab.binding.withings.internal.service.AccessTokenService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * @author Sven Strohschein - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractAPIHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractAPIHandler.class);
+
+    private final AccessTokenService accessTokenService;
+    private final HttpClient httpClient;
+
+    public AbstractAPIHandler(AccessTokenService accessTokenService, HttpClient httpClient) {
+        this.accessTokenService = accessTokenService;
+        this.httpClient = httpClient;
+    }
+
+    protected <@NonNull T extends BaseResponse> Optional<T> executeAuthPOSTRequest(String url, String action,
+            Map<String, String> parameters, Class<T> responseType) {
+        return executePOSTRequest(url, action, parameters, responseType, false);
+    }
+
+    protected <@NonNull T extends BaseResponse> Optional<T> executePOSTRequest(String url, String action,
+            Map<String, String> parameters, Class<T> responseType) {
+        return executePOSTRequest(url, action, parameters, responseType, true);
+    }
+
+    protected <@NonNull T extends BaseResponse> Optional<T> executePOSTRequest(String url, String action,
+            Map<String, String> parameters, Class<T> responseType, boolean useAccessToken) {
+
+        Optional<String> accessToken;
+        if (useAccessToken) {
+            accessToken = accessTokenService.getAccessToken();
+            if (!accessToken.isPresent()) {
+                logger.warn("No access token available for the access to the Withings data API!");
+                return Optional.empty();
+            }
+        } else {
+            accessToken = Optional.empty();
+        }
+
+        Fields fields = new Fields(true);
+        fields.put("action", action);
+        for (Map.Entry<String, String> parameterMapEntry : parameters.entrySet()) {
+            fields.put(parameterMapEntry.getKey(), parameterMapEntry.getValue());
+        }
+
+        Request request = httpClient.POST(url).timeout(20, TimeUnit.SECONDS).idleTimeout(60, TimeUnit.SECONDS)
```

Comment:
Good question, but it can make a difference. ;-) At least it can affect web sockets were the connection is still open when a response is sent. I learned that with issue https://github.com/openhab/openhab-addons/issues/8734